### PR TITLE
feat: tags replaces GrantCategory

### PR DESCRIPTION
**Pull request by thelostone-mc**
##### Description

This PR does a few things

- removed the GrantCategory link to GrantType
- introduced a new table GrantTag (which is to eventually replace GrantCategory)
- has a custom migration to 
 - port GrantCategory to Grant Tag
 - port Grant -> GrantCategory mapping to Grant -> GrantTag mapping
- Document the actual cards_info endpoint to ensure it's easy to navigate
- replace category with grant_tag on the backend 
- adds GrantTag to admin and removed GrantCategory
- updates the cares_info endpoint to enable filtering by 
  - grant_tag `/grants/cards_info?grant_tag=lol`
  - tenants `/grants/cards_info?tenants=eth,polkadot` // this is case insensitive 
- introduces to new endpoint `http://localhost:8000/grants/v1/api/tags?q=` to query and get tags
- Also ensures MatchPledge to work with GrantTag

**What is not handled in this PR**

- **Grant Creation is blocked as GrantTags would have to be loaded**
- Removing of GrantCategory and removing Grant.categories (this will be done later in a follow up PR once we are sure data is migrated right and verified on live)
- Right now, the tags follow a simple icontains search. This is cause it's too early for us to implement anything right now and the understanding is for GR11 we restrict the tags to a finite set. As this grows -> we'd explore using SearchVector / haystack / trie 

**Existing API changes**

- You can search for grants in multiple grant_types
`http://localhost:8000/grants/cards_info?idle=true&grant_types=media,matic`
- You can search for grants in multiple grant_types
`http://localhost:8000/grants/cards_info?idle=true&grant_tags=L1,L2`
- You can search for grants in multiple tenants
`http://localhost:8000/grants/cards_info?idle=true&tenants=eth,polkadot`
- within `cards_info`, the following code has been change
 - Retired param + code
   - sort param
   - following grants
   - only_contributions
 - Renamed param
   - type -> grant_types
- other pieces retired to make backend quicker
  - remove match pledge call (current/past partners)
  - background for grants page call

### CODE REVIEW

File: app/grants/views.py [hunk 1]
```diff
@@ -570,80 +577,103 @@ def get_grants(request):
...
     if not idle_grants:
+        # 7. Filter grants which are stale (last update was > 3 months )
         _grants = _grants.filter(last_update__gt=three_months_ago)
 
     if state == 'active':
+        # 8. Filter grants which are active
         _grants = _grants.active()
 
-    if grant_type != 'all' and grant_type != 'me' and grant_type != 'collections':
+    if grant_type not in ['all', 'me', 'collectiions']:
```
Comment (octavioamu) [inline app/grants/views.py:640]: is a typo? 'collectiions'

File: app/grants/views.py [hunk 2]
```diff
@@ -997,10 +1007,23 @@ def grants_by_grant_type(request, grant_type):
 
 
     response = TemplateResponse(request, 'grants/index.html', params)
-    response['X-Frame-Options'] = 'SAMEORIGIN'
+    response['X-Frame-Optionswh'] = 'SAMEORIGIN'
```
Comment (octavioamu) [inline app/grants/views.py:1010]: is this a valida header o was changed by mistake?

File: app/grants/models.py [hunk 3]
```diff
@@ -553,7 +567,8 @@ class Meta:
         null=True,
         blank=True,
     )
-    categories = models.ManyToManyField(GrantCategory, blank=True)
+    categories = models.ManyToManyField(GrantCategory, blank=True) # TODO: REMOVE
+    tags = models.ManyToManyField(GrantTag, blank=True)
```
Comment (octavioamu) [inline app/grants/models.py:571]: I think there is a confusion here tags doesn't relate to types, could be any combination of type / tag as the users will create those

**Comment by octavioamu** [comment]
seems there is a problem with the clr links 
going to something like `http://localhost:8000/grants/clr/ethereum/10/gr10-gitcoindao` 
UnboundLocalError: local variable 'category' referenced before assignment

File "/code/app/grants/views.py", line 424, in clr_grants
return grants_by_grant_clr(request, clr_round)
File "/code/app/grants/views.py", line 1152, in grants_by_grant_clr
'selected_category': category,

**Comment by thelostone-mc** [comment]
@octavioamu fixed all those ! have a look at the description and you'll notice a few other changes made to the API

File: app/grants/views.py [hunk 4]
```diff
@@ -365,15 +370,15 @@ def get_collections(user, keyword, sort='-shuffle_rank', collection_id=None, fol
 
 
 def bulk_grants_for_cart(request):
-    grant_type = request.GET.get('type', 'all')
-    sort = request.GET.get('sort_option', 'weighted_shuffle')
+    grant_types = request.GET.get('type', None)
+    # sort = request.GET.get('sort_option', 'weighted_shuffle')
```
Comment (gdixon) [inline app/grants/views.py:374]: Are we removing the sort altogether?

File: app/grants/views.py [hunk 5]
```diff
@@ -1053,18 +1091,18 @@ def grants_by_grant_clr(request, clr_round):
...
+    #     current_partners_fund += partner.amount
 
     grant_types = get_grant_clr_types(clr_round, network=network)
 
-    grants_following = Favorite.objects.none()
+    # grants_following = Favorite.objects.none()
     collections = []
     if request.user.is_authenticated:
-        grants_following = Favorite.objects.filter(user=request.user, activity=None).count()
+        # grants_following = Favorite.objects.filter(user=request.user, activity=None).count()
```
Comment (gdixon) [inline app/grants/views.py:1105]: Following/Favourite is still an option in the "me menu" - do we want to be removing this?

File: app/grants/views.py [hunk 6]
```diff
@@ -342,9 +347,9 @@ def get_collections(user, keyword, sort='-shuffle_rank', collection_id=None, fol
     if idle_grants:
         _collections = _collections.filter(grants__last_update__gt=three_months_ago)
 
-    if only_contributions:
-        contributions = user.profile.grant_contributor.filter(subscription_contribution__success=True).values('grant_id')
-        _collections = _collections.filter(grants__in=Subquery(contributions))
+    # if only_contributions:
```
Comment (thelostone-mc) [inline app/grants/views.py:350]: uncomment . will be used in me menu

File: app/grants/views.py (hunk 4 above)
Comment (thelostone-mc) [reply app/grants/views.py:374]: uncomment . will be used in me menu

File: app/grants/views.py [hunk 7]
```diff
@@ -365,15 +370,15 @@ def get_collections(user, keyword, sort='-shuffle_rank', collection_id=None, fol
...
+    # sort = request.GET.get('sort_option', 'weighted_shuffle')
     network = request.GET.get('network', 'mainnet')
     keyword = request.GET.get('keyword', '')
     state = request.GET.get('state', 'active')
-    category = request.GET.get('category', '')
+    grant_tags = request.GET.get('grant_tags', '')
     idle_grants = request.GET.get('idle', '') == 'true'
-    following = request.GET.get('following', '') != ''
-    only_contributions = request.GET.get('only_contributions', '') == 'true'
+    # following = request.GET.get('following', '') != ''
```
Comment (thelostone-mc) [inline app/grants/views.py:380]: uncomment . will be used in me menu

File: app/grants/views.py [hunk 8]
```diff
@@ -554,96 +567,135 @@ def get_grants(request):
...
                         list(profile.grant_admin.all().values_list('pk', flat=True))
             _grants = _grants.exclude(id__in=grants_id)
-        elif grant_type == 'me':
-            grants_id = list(profile.grant_teams.all().values_list('pk', flat=True)) + \
-                        list(profile.grant_admin.all().values_list('pk', flat=True))
-            _grants = _grants.filter(id__in=grants_id)
-        elif only_contributions:
-            contributions = profile.grant_contributor.filter(subscription_contribution__success=True).values('grant_id')
-            _grants = _grants.filter(id__in=Subquery(contributions))
+        # elif grant_types == 'me':
```
Comment (thelostone-mc) [inline app/grants/views.py:625]: uncomment . will be used in me menu

File: app/grants/views.py [hunk 9]
```diff
@@ -554,96 +567,135 @@ def get_grants(request):
...
+        #     # 4. Filter grants created by user
+        #     grants_id = list(profile.grant_teams.all().values_list('pk', flat=True)) + \
+        #                 list(profile.grant_admin.all().values_list('pk', flat=True))
+        #     _grants = _grants.filter(id__in=grants_id)
+        # elif only_contributions:
+        #     # 5. Filter grants to which the user has contributed to
```
Comment (thelostone-mc) [inline app/grants/views.py:631]: uncomment . will be used in me menu

File: app/grants/views.py [hunk 10]
```diff
@@ -554,96 +567,135 @@ def get_grants(request):
...
+    if grant_types and grant_types not in ['all', 'collections']:
+        # 9. Fetch grants which have mentioned grant_types
+        types= grant_types.split(',')
+        type_query= Q()
+        for grant_type in types:
+            type_query |= Q(grant_type__name=grant_type)
+
+        _grants = _grants.filter(type_query)
+
+    # if following and request.user.is_authenticated:
```
Comment (thelostone-mc) [inline app/grants/views.py:657]: uncomment . will be used in me menu

File: app/grants/views.py [hunk 11]
```diff
@@ -554,96 +567,135 @@ def get_grants(request):
...
+                tenant = tenant.lower()
+                tenant_filter = tenant + '_payout_address' if tenant != 'eth' else 'admin_address'
+                tenant_query |= Q(tenant_filter!='0x0')
+
+            # _grants = _grants.exclude(Q(**{tenant_filter: '0x0'}))
+        _grants = _grants.filter(tenant_query)
+
+
+    # # 13. Sort filtered grants
+    # if 'match_pledge_amount_' in sort:
```
Comment (thelostone-mc) [inline app/grants/views.py:685]: uncomment . will be used in me menu

**Comment by thelostone-mc** [comment]
closing this out as this will be merged in @octavioamu PR

**Pull request closed without merging**